File: 029_divide_two_integers/Solution.go
package main

import (
	"fmt"
	"math"
)

func divide(dividend int, divisor int) int {
	sign := 1
	if (dividend > 0) != (divisor > 0) {
		sign = -1
	}
	m, n := abs(dividend), abs(divisor)
	res := 0

	for m >= n {
		t, p := n, 1
		for m > (t << 1) {
			t <<= 1
			p <<= 1
		}
		m -= t
		res += p
	}

	if sign == 1 {
		return int(math.Min(float64(res), math.MaxInt32))
	} else {
		return int(math.Max(float64(-res), math.MinInt32))
	}
}

func abs(x int) int64 {
	if x > 0 {
		return int64(x)
	} else {
		return -int64(x)
	}
}

func main() {
	fmt.Println(divide(-2147483648, 1))
}

File: 029_divide_two_integers/Solution.py
class Solution:
    def divide(self, dividend: int, divisor: int) -> int:
        res = 0
        sign = -1 if (dividend > 0) ^ (divisor > 0) else 1
        dividend, divisor = abs(dividend), abs(divisor)
        while dividend >= divisor:
            t, p = divisor, 1
            while dividend > (t << 1):
                t <<= 1
                p <<= 1
            dividend -= t
            res += p

        if sign == 1: return min(res, (1 << 31) - 1)
        else: return max(-res, - 1 << 31)


if __name__ == '__main__':
    solution = Solution()
    print(solution.divide(-2147483648, -1))
